File: 238-product-of-array-except-self/238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    //2 iterations, once to establish prefixes, then iterate from the right, multiply by postfix product
    let prefixProduct = 1;
    let result = [];
    for (let i = 0; i < nums.length; i++) {
        result.push(prefixProduct)
        prefixProduct *= nums[i];    
    }
    
    let postfixProduct = 1;
    for (let i = nums.length-1; i >= 0; i--) {
        result[i] = postfixProduct * result[i]
        postfixProduct *= nums[i]
    }
    return result
};

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution: # 87.32%, 89.5%
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # leftProduct as 1
        # rightProduct as 1
        # res array

        leftProd = 1
        rightProd = 1
        res = len(nums)*[0]

        # 1st loop: populate res with left products. populate res first then update leftProd
        for i, num in enumerate(nums):
            res[i] = leftProd
            leftProd *= num

        # 2nd loop: iterate backwards. multiply left prods with right prods. set then update rightProd
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= rightProd
            rightProd *= nums[i]

        return res
